### PR TITLE
o/snapstate: implicitly set --unaliased flag for parallel installs

### PR DESCRIPTION
**Pull request by ZeyadYasser**
Implicitly set `--unaliased` flag for parallel installs to avoid alias conflicts with the main snap.

Some potential improvement points:
- Extracting parallel install detection into a function to make it more explicit instead of doing `InstanceKey != ""` implicitly
- Adding a unit test for installing many snaps at once with parallel installs included

They seemed a bit out of scope for this PR, if they seem sensible, maybe they can be implemented in their own PRs?

### CODE REVIEW

File: tests/main/parallel-install-auto-aliases/task.yaml [hunk 1]
```diff
@@ -73,14 +69,21 @@ execute: |
...
+    echo "Even when test-snapd-auto-aliases_foo is installed first"
+    echo "it should be installed with --unaliased"
     snap install test-snapd-auto-aliases_foo
 
-    echo "Installing test-snapd-auto-aliases will conflict"
-    # TODO parallel-install: need to install using file
-    not snap install "$name"
-    snap change --last=install | MATCH 'cannot enable aliases .* for "test-snapd-auto-aliases", already enabled for "test-snapd-auto-aliases_foo"'
+    echo "Installing test-snapd-auto-aliases after test-snapd-auto-aliases_foo"
+    echo "shouldn't conflict and it's aliases should be enabled"
```
Comment (MiguelPires) [inline tests/main/parallel-install-auto-aliases/task.yaml:77]:
```suggestion
    echo "shouldn't conflict and its aliases should be enabled"
```

File: tests/main/parallel-install-auto-aliases/task.yaml [hunk 2]
```diff
@@ -73,14 +69,21 @@ execute: |
...
+    snap install test-snapd-auto-aliases
```
Comment (MiguelPires) [inline tests/main/parallel-install-auto-aliases/task.yaml:78]: I guess this check is time dependent right? If it was failing, it's possible that we wouldn't know before the first install finished before we started this one

**Review by MiguelPires — COMMENTED** [review]
Some small comments

File: tests/main/parallel-install-auto-aliases/task.yaml (hunk 2 above)
Comment (MiguelPires) [reply tests/main/parallel-install-auto-aliases/task.yaml:78]: Oh sorry never mind, I misread the test

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -365,6 +365,13 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		}
 	}
 
+	// Implicitly set --unaliased flag for parallel installs to avoid
+	// alias conflicts with the main snap
+	if snapsup.InstanceKey != "" {
+		logger.Noticef("implicitly setting --unaliased flag for parallel install %q", snapsup.InstanceName())
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply overlord/snapstate/snapstate.go]: Updated, Thanks!

File: tests/main/parallel-install-auto-aliases/task.yaml (hunk 1 above)
Comment (ZeyadYasser) [reply tests/main/parallel-install-auto-aliases/task.yaml]: Updated, Thanks!

File: overlord/snapstate/snapstate_update_test.go [hunk 4]
```diff
@@ -1408,6 +1408,8 @@ func (s *snapmgrTestSuite) TestParallelInstanceUpdateRunThrough(c *C) {
 		Version:     "services-snapVer",
 		PlugsOnly:   true,
 		InstanceKey: "instance",
+		// parallel installs should implicitly pass --unaliased
+		Flags: snapstate.Flags{Unaliased: true},
```
Comment (pedronis) [inline overlord/snapstate/snapstate_update_test.go:1412]: this shouldn't be required as refreshes should preserve the current aliases setup?

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -365,6 +365,12 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		}
 	}
 
+	// Implicitly set --unaliased flag for parallel installs to avoid
+	// alias conflicts with the main snap
+	if snapsup.InstanceKey != "" {
+		snapsup.Unaliased = true
+	}
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:372]: as we mainly we want to affect installs, as refreshes should follow the current state of things, we usually put this kind of logic in ensureInstallPreconditions I think

**Review by pedronis — CHANGES_REQUESTED** [review]
wondering about logic placement and effect/need on refreshes

File: overlord/snapstate/snapstate.go (hunk 5 above)
Comment (ZeyadYasser) [reply overlord/snapstate/snapstate.go]: Thanks for pointing this out, Updated!

File: overlord/snapstate/snapstate_update_test.go (hunk 4 above)
Comment (ZeyadYasser) [reply overlord/snapstate/snapstate_update_test.go]: Yes, thank you for pointing this out, updated!

**Review by mvo5 — APPROVED** [review]
Thank you!

**Pull request merged**